website: Add dark theme

thesuperzapper opened the pull request:
This PR adds a new dark-theme to the Kubeflow website.

resolves https://github.com/kubeflow/website/issues/3692
resolves https://github.com/kubeflow/website/issues/3954

# How does this work?

Because we use docsy, which has its own themes in addition to bootstrap, we now render two themes into our CSS based on the `[data-theme]` attribute on the root `<html>` element, see `assets/scss/main.scss` for more context.

This approach allows us to retain our custom docsy changes while adding a nice dark theme.

# What changed?

__NOTE:__ this PR does NOT change any of the docs content. However, some images/diagrams have been updated to look better when viewed on a dark background.

## Theme-specific styling

To make theme-specific styling possible, we have added the following:

- __CSS classes:__
    - `show_light-only` (only show element in light mode)
    - `show_dark-only` (only show element in dark mode)
- __Theme-specific SCSS variables:__
    - `assets/scss/dark/_variables.scss`
    - `assets/scss/light/_variables.scss`
- __Theme-specific SCSS styles:__
    - `assets/scss/dark/_styles.scss`
    - `assets/scss/light/_styles.scss`

__NOTE:__ the existing `assets/scss/_styles_project.scss` is still used for global styles. It is used by both themes, usually with theme-specific variables to set colors.

## Theme Switcher

We have added a theme switcher to the top right of the website, which allows users to switch between the light and dark themes. By default, we match the theme set on the user's OS.

![theme_switcher](https://github.com/user-attachments/assets/51406445-1bd6-497d-a2af-d688aff3ce08)

## Images

We have not created new images for the dark theme, as most looked OK on the dark background. In the future, we can use the `show_light-only` and `show_dark-only` classes to show different images based on the theme.

We have updated the following images to look better on the dark theme:

- Getting Started / Introduction
- ~Getting Started / Architecture~ (split off into https://github.com/kubeflow/website/pull/4038 and https://github.com/kubeflow/website/pull/4068)

## Iframes / Swagger

There were two cases where I used a color inversion filter to create a fake "dark mode":

1. For the Google calendar embed in the community page
2. For the Swagger UI API reference pages

They seem to work well enough, but we can revisit this in the future if those components get official dark mode support.

# Screenshots

## Home Page ([preview](https://deploy-preview-3981--competent-brattain-de2d6d.netlify.app/))

![Home](https://github.com/user-attachments/assets/740d58a5-d06d-46ee-b17a-262cbd01d27c)

## Getting Started ([preview](https://deploy-preview-3981--competent-brattain-de2d6d.netlify.app/docs/started/))

![Getting Started](https://github.com/user-attachments/assets/ea4349ed-32f5-4518-be2d-079d7989f432)

## Code Blocks ([preview](https://deploy-preview-3981--competent-brattain-de2d6d.netlify.app/docs/components/pipelines/getting-started/))

![Code Blocks](https://github.com/user-attachments/assets/7628f789-66d7-4de2-ab98-17a0ac90c73d)

thesuperzapper commented:
@varodrig @ederign @milinddethe15 could you take a first look at this PR, that adds a dark mode?

Once we are happy, we can share it around the community for wider feedback.

## See: [preview site](https://deploy-preview-3981--competent-brattain-de2d6d.netlify.app/)

sftim reviewed (commented):
A trick I used for Kubernetes: one logo, change the color using CSS. You have to embed the SVG into the HTML for this to work, but it does work nicely.

I used it to change the text of the _kubernetes_ text in the logo depending on the background. Change is not merged but I can show an example if you're not sure what I'm on about.

andreyvelich commented on assets/icons/logo-black.svg:
We need to make sure that these logos are in sync with CNCF artwork
https://github.com/cncf/artwork/tree/main/projects/kubeflow/icon

cc @nate-double-u @krook

andreyvelich commented on content/en/docs/started/images/kubeflow-architecture.drawio.svg:
A few points for the diagram:
1. We should keep KServe in the Kubeflow Components diagram since it is part of Kubeflow Ecosystem as we discussed many time before. Also, it presents at this diagram: https://www.kubeflow.org/docs/started/introduction/#kubeflow-overview-diagram 
2. Why do you prefer Chips over Hardware ?
3. What do we think about removing the External Add-Ons section from this diagram ?
I think, we discussed it many times before that it might be hard to keep it up-to-date. So we can just have 1 single page where other projects can link themself: https://github.com/kubeflow/website/issues/3441#issuecomment-2651660453 

cc @kubeflow/wg-pipeline-leads @kubeflow/wg-training-leads @kubeflow/wg-data-leads @kubeflow/wg-automl-leads @akgraner @chasecadet @kubeflow/kubeflow-steering-committee @franciscojavierarceo @kubeflow/wg-manifests-leads @kubeflow/wg-training-leads @varodrig @kubeflow/wg-notebooks-leads @rimolive @hbelmiro for review.

thesuperzapper replied:
Lets discuss this in https://github.com/kubeflow/website/pull/4038

varodrig commented on content/en/docs/components/pipelines/legacy-v1/introduction.md:
```diff
@@ -209,25 +209,25 @@ Prediction results:
 
 ## Architectural overview
 
 <img src="/docs/images/pipelines-architecture.png" 
```
![pipelines-dark](https://github.com/user-attachments/assets/9e6633ac-400d-43ea-b078-5798cace2829)
In the diagram there are few words that are not visible due to the dark background.

thesuperzapper replied:
This has been fixed.

varodrig reviewed (commented):
Amazing work! thank you for doing this.

I reviewed all the pages and only have a few comments 

1-  is that when the source code is commented the colour is grey and it's not very clear when the background is dark. Here's an image to see how it looks.
Compile a Pipeline 
docs/components/pipelines/user-guides/core-functions/compile-a-pipeline/

2- The GitHub Logo is not very visible
docs/components/pipelines/legacy-v1/sdk/python-function-components/

![image](https://github.com/user-attachments/assets/3244b84f-a36a-4ede-9ae6-be90e532b4cc)


![image](https://github.com/user-attachments/assets/59c8110a-b6ca-421a-9c46-868bc2fa4ff8)

consider 
https://brand.github.com/foundations/logo

3- the webpage docs/components/pipelines/legacy-v1/sdk/pipelines-with-tekton/ is not showing the source code instead a few white blocks.

![image](https://github.com/user-attachments/assets/ddc746f5-a0b2-486a-9e63-d57eaeeab201)

4- model registy diagram
docs/components/model-registry/overview/
the lines and letters are not very visible.
![image](https://github.com/user-attachments/assets/1b5f8195-33f6-4cab-be7a-b4e4f98f50a9)
5- I left another comment on the code.

franciscojavierarceo commented:
Thanks for diligence here @varodrig!

@thesuperzapper the dark mode feature is a significant improvement for those with eye sight sensitivity (I am one of those people so I deeply appreciate this). Some of our diagrams and logos don't render well, and that's not great, but I think we can chip away at those overtime in follow up PRs. 

I don't think we need to block releasing this update until the website is fully fixed. WDYT?

varodrig commented:
@franciscojavierarceo  @juliusvonkohout   @andreyvelich  can you approve?
Thanks 

cc @thesuperzapper

franciscojavierarceo commented:
looks like the conflicts have to be resolved

/approve

thesuperzapper commented:
@franciscojavierarceo @varodrig will rebase in the next day or so, if we are all happy with the style proposed.

varodrig commented:
> looks like the conflicts have to be resolved
> 
> /approve

thank you

varodrig commented:
> @franciscojavierarceo @varodrig will rebase in the next day or so, if we are all happy with the style proposed.

Sounds great! thank you for your work on this

andreyvelich commented:
/hold
We should not merge this until we discuss the architecture diagram as I mentioned here: https://github.com/kubeflow/website/pull/3981#discussion_r1953463244

franciscojavierarceo commented:
That's fine @andreyvelich 

I did state explicitly that I think this was still a net benefit all else equal (i.e., the scope of the architecture change should probably be outside of this PR).

See [this comment](https://github.com/kubeflow/website/pull/3981#issuecomment-2665645588):

>@thesuperzapper the dark mode feature is a significant improvement for those with eye sight sensitivity (I am one of those people so I deeply appreciate this). Some of our diagrams and logos don't render well, and that's not great, but I think we can chip away at those overtime in follow up PRs.
>
>I don't think we need to block releasing this update until the website is fully fixed. WDYT?

andreyvelich commented:
>I did state explicitly that I think this was still a net benefit all else equal (i.e., the scope of the architecture diagram change should probably be outside of this PR).

In that case, can we remove the diagram update from this PR and work on it as a followup ?

varodrig commented:
> > n we remove the diagram update from this PR and work on it as a followup ?
> 
> +1

@thesuperzapper  can we make the updates to merge this PR?
thank you

thesuperzapper commented:
@andreyvelich @franciscojavierarceo the dark-mode PR should be ready to merge now, I have rebased and split the ecosystem diagram updates into a separate PR: https://github.com/kubeflow/website/pull/4038.

I have also addressed most of the issues found by @varodrig in https://github.com/kubeflow/website/pull/3981#pullrequestreview-2622462936.

__WARNING:__ merging the PR will be a significant change for many users, so I would like both @andreyvelich and @franciscojavierarceo to say they are happy before we merge.

andreyvelich commented on content/en/_index.html:
```diff
@@ -3,26 +3,31 @@
 +++
```
What is the default mode for the website or it auto-detects the browser settings ?

thesuperzapper replied:
It follows the computers dark/light mode by default, or uses light if it cant detect.

andreyvelich commented on content/en/_index.html:
```diff
@@ -3,26 +3,31 @@
-  <h3 class="section-head">What is Kubeflow?</h3>
   <div class="container">
+    <h3 class="section-head">
+      <span class="border-bottom pb-2">What is Kubeflow?</span>
```
Why did we change the font color in the light mode ?

thesuperzapper replied:
The darker blue for the headings is easier to read on the white background, @ederign can confirm if he agrees?

andreyvelich replied:
But this goes against the style for other pages, isn't ?
<img width="578" alt="Screenshot 2025-03-11 at 20 11 19" src="https://github.com/user-attachments/assets/f5fab5f4-aa99-4b42-bc35-20766bc6db37" />

thesuperzapper replied:
Thats actually another important reason to use the darker blue, because the lighter color is used for links, and those headings are not links.

andreyvelich replied:
I feel like the text font should be at least aligned with the Kubeflow hex colors: https://github.com/kubeflow/marketing-materials/blob/master/logos/Logo-File-Guide.pdf
```
#4279F4
#0028AA
```

Right now, it has this hex:
```
#223C7A
```

thesuperzapper replied:
The "official" dark blue (`#0028AA`) looks really bad in most contexts, including this one.

The alternative dark blue (`#213d7a`) which has been used all across the website for the past 7+ years looks much nicer, I have given it a special hex code in https://github.com/kubeflow/website/pull/3981/commits/48e9ffff2673527023e377ce4036a719f12eeb2b so we understand its not a mistake.

andreyvelich commented on content/en/_index.html:
```diff
@@ -84,35 +91,34 @@ <h5 class="card-title text-white section-head">Notebooks</h5>
+        <div class="card-body bg-primary-dark">
+          <div class="card-title">Model Training</div>
+          <p class="card-text">
+            <a href="/docs/components/trainer/overview/" target="_blank" rel="noopener">Kubeflow Trainer</a> is a Kubernetes-native interface for training models and fine-tuning LLMs.
```
Can we keep the original description, since we need to align with @kubeflow/wg-training-leads @Electronic-Waste @astefanutti for it.

```suggestion
            <a href="/docs/components/trainer/overview/" target="_blank" rel="noopener">Kubeflow Trainer</a> is a Kubernetes-native project
             designed for LLMs fine-tuning and enabling scalable, distributed training of ML models across
             various frameworks, including PyTorch, JAX, TensorFlow, and others.```

andreyvelich replied:
I am ok to splint this sentence between 2, but we should say this:
> Kubeflow Trainer is a Kubernetes-native project for LLMs fine-tuning and ML models training. It enables scalable and distributed training with various frameworks including PyTorch, JAX, TensorFlow, and more.

Any other thoughts @kubeflow/wg-training-leads ?

thesuperzapper replied:
"LLMs fine-tuning" is not a valid english, the correct way to say it would be "LLM fine-tuning".

Also, I think having "training models" first is important, because training operator is not only about LLMs, and having it first might send the wrong message to new users.

What about this:

> Kubeflow Trainer is a Kubernetes-native tool for training models and fine-tuning LLMs. It enables scalable and distributed training with various frameworks including PyTorch, JAX, TensorFlow, and more.

Or if you really want to use "project":

> Kubeflow Trainer is a Kubernetes-native project for training models and fine-tuning LLMs. It enables scalable and distributed training with various frameworks including PyTorch, JAX, TensorFlow, and more.

andreyvelich replied:
>"LLMs fine-tuning" is not a valid english, the correct way to say it would be "LLM fine-tuning".

LLM fine-tuning also works, I can see that HF does the same: https://huggingface.co/docs/autotrain/en/llm_finetuning

>Also, I think having "training models" first is important, because training operator is not only about LLMs, and having it first might send the wrong message to new users.

In the current stage of ML and GenAI, majority of ML Users just do post-training (e.g. fine-tuning), so this capability is more important then saying that is enables training ML models.

andreyvelich replied:
>Or if you really want to use "project":

I would like to use project, since we want to rename Kubeflow Components to Kubeflow Projects as part of: https://github.com/kubeflow/website/pull/4023.

thesuperzapper replied:
Resolved in https://github.com/kubeflow/website/pull/3981/commits/b135fbd6639cfa788dbbc8b83cfb6f5261cff0fa

Copy is now:

> Kubeflow Trainer is a Kubernetes-native project for fine-tuning LLMs and model training.
> It enables scalable and distributed training with various frameworks including PyTorch, JAX, TensorFlow, and more.

andreyvelich commented on content/en/_index.html:
```diff
@@ -84,35 +91,34 @@ <h5 class="card-title text-white section-head">Notebooks</h5>
+     <div class="card">
+        <a href="/docs/components/trainer/overview/" target="_blank" rel="noopener" >
+          <img
+            src="/docs/images/logos/trainer.png"
```
Can you keep the original logo with Kubeflow and Trainer word since it is the official Logo: https://github.com/kubeflow/trainer?tab=readme-ov-file#kubeflow-trainer

thesuperzapper replied:
From a design perspective, I think we should avoid text in those images.

Also, the image needs to be square (for consistent card sizes), so the logo would need to be smaller if we keep the text. If you are really against using the icon I will make a square version with the text, but I think it will look worse.

If it helps, don't think of it as a "logo" but "icon".

### OLD / NEW

<img width="25%" alt="Screenshot 2025-03-11 at 13 10 54" src="https://github.com/user-attachments/assets/fbeb1e0d-de2c-40fd-95d6-5ca3a6c18b8f" /> / <img width="25%" alt="Screenshot 2025-03-11 at 13 12 13" src="https://github.com/user-attachments/assets/f72e2cca-2125-4479-b846-d85876baa3d5" />

andreyvelich replied:
> If you are really against using the icon I will make a square version with the text, but I think it will look worse.
If it helps, don't think of it as a "logo" but "icon".

Please update the logo to a square version while keeping the name included. Maintaining the 'Kubeflow Trainer' branding is important, as the logo alone may not be recognizable to new users.


cc @franciscojavierarceo @kubeflow/wg-training-leads @astefanutti @Electronic-Waste for feedback.

tenzen-y replied:
I think we can pick either art from https://github.com/cncf/artwork/blob/main/examples/incubating.md#kubeflow-trainer-logos

andreyvelich replied:
That's true, but I suggest selecting artwork that includes naming, as the logo and project name are new, and we want to increase exposure for the Kubeflow Trainer brand.

tenzen-y replied:
I do not have any preference here. I'm ok in case of the CNCF registered logo is selected.

franciscojavierarceo replied:
+1 to keeping name

thesuperzapper replied:
If we keep the name, we will need to make the Trainer logo smaller so it fits in a 1:1 ratio square (otherwise it misaligns the cards).

Personally, I think the icon looks great on its own (and the icon is listed as an official logo [here](https://github.com/cncf/artwork/blob/main/examples/incubating.md#kubeflow-trainer-logos)), but I will defer to others with more UX experience.

@ederign do you have an opinion?

andreyvelich replied:
We can request CNCF to provide additional logo samples and adjust its size accordingly.
I believe the decision to retain the project name in the logo should be made by the Kubeflow Trainer maintainer.

If you’d like to move forward with this PR, you can revert the logo change from this PR.

ederign replied:
First, we need to have a 1:1 ratio logo for Kubeflow Trainer. But we can deal with this in a separate PR. @andreyvelich, how do we request those logos? I can reach out to them.

About having the project name on the logo or not, I agree with @andreyvelich that this is up to the project contributors.

On the other hand, I agree with @thesuperzapper that our home page would look better if we can have just icons on that, mainly because there is a 'caption' box explaining each project above each icon.

Can we have for each project, besides their logo (where they can decide whatever they want on it), also an icon for cases like the home page where it will be used with a 'caption'?  (an icon will also look great on small images)

@thesuperzapper probably we will need in this case a icon also for Notebooks.

andreyvelich replied:
I created, CNCF Service Desk ticket: https://cncfservicedesk.atlassian.net/servicedesk/customer/portal/1/CNCFSD-2728

thesuperzapper replied:
I have used the logos with words in them after https://github.com/kubeflow/website/pull/3981/commits/180f2c8cee262fb4d129c5e6f422666144a33c2c

To make them all fit within a 1:1 image, I have simply added padding, which looks ok.

andreyvelich commented on assets/scss/dark/_variables.scss:
```diff
@@ -0,0 +1,51 @@
+// --------------------------------------------------
+// Search
+// --------------------------------------------------
+$search-icon-color: $teal;
+$search-placeholder-color: darken($teal, 10%);
+$search-border-color: darken($teal, 20%);
```
For me the search bar right now green even in light mode:

<img width="772" alt="Screenshot 2025-03-11 at 19 42 56" src="https://github.com/user-attachments/assets/cab0f61a-2c4c-451a-9e7c-ab3565067560" />

thesuperzapper replied:
The idea was to highlight it because many users miss the fact we even have a search bar.

I will defer to @ederign if he thinks it looks better/worse.

ederign replied:
@thesuperzapper I understand that the search is kind of hidden, but I would not use a green color here.

For the dark theme, I believe if we put the border, icon, and text in white, it would have a similar effect. We can also increase the border for 2px to highlight it.

thesuperzapper replied:
I have updated the search styles in https://github.com/kubeflow/website/pull/3981/commits/e2d8a1cb9c6ab5588cbab857ae99e561e8eaaf9c so the green border is only visible when the search input is in focus.

I used a drop-shadow to highlight the presence of the search bar, so people are less likely to miss it.

> <img width="265" alt="Screenshot 2025-03-29 at 17 02 34" src="https://github.com/user-attachments/assets/d86c36e3-951b-4386-8e80-cdc1e9f6d25d" />

> <img width="271" alt="Screenshot 2025-03-29 at 17 02 28" src="https://github.com/user-attachments/assets/0a9d79aa-7be7-422c-a56a-ed43c4cf0f8a" />

andreyvelich commented on content/en/docs/images/logos/kserve.png:
I am not sure if we should add the KServe logo without the name in the frontpage.
cc @terrytangyuan for the feedback

thesuperzapper replied:
My perspective is the same as for Kubeflow Trainer, that we should use the "icons" rather than the logos, so we can avoid text in the images.

franciscojavierarceo replied:
I don't think people will remember the logos, especially if we want to more users then, by definition, most won't know the logo.

thesuperzapper replied:
I have used the logos with words in them after https://github.com/kubeflow/website/pull/3981/commits/180f2c8cee262fb4d129c5e6f422666144a33c2c for Trainer, Katib, and KServe.

To make them all fit within a 1:1 image, I have simply added padding, which looks ok.

andreyvelich reviewed (commented):
I left my initial questions.
It would be nice to get someone from the UI/UX team to review these changes.
cc @ederign @StefanoFioravanzo

andreyvelich commented on content/en/docs/started/images/ml-lifecycle-kubeflow.drawio.svg:
Why did you change Feature Extraction step in this diagram?
It should go from KServe to Online Feature store, isn't @franciscojavierarceo ?

andreyvelich replied:
Also, if you are already making changes in this diagram, please can you keep Training Step before Optimization Step and keep boxes updated for Kubeflow Trainer and Kubeflow Katib

![ml-lifecycle-kubeflow drawio](https://github.com/user-attachments/assets/2661331a-537f-4781-b084-da82664a03f8)

thesuperzapper replied:
Ahh, that was a "typo" of sorts, I was making the diagram more "square" (and fixing the Kubeflow Logo) but accidentally moved the connecting lines on the "feature extraction" node.

thesuperzapper replied:
I have reverted all changes to these diagrams in https://github.com/kubeflow/website/pull/3981/commits/31ae79f0576239ce1ae17b3c3ae41d056db4b7a0.

I will raise a separate PR which updates them so we can discuss.

thesuperzapper commented:
@andreyvelich I have resolved all pending comments, and split out the diagram updates into separate PRs so they don't block us from merging (https://github.com/kubeflow/website/pull/4038, _other one TBA_).

I think we should be good to merge this now, right before Kubecon!

andreyvelich commented on content/en/_index.html:
```diff
@@ -36,10 +41,12 @@ <h3 class="section-head">What is Kubeflow?</h3>
 <section id="pageContent">
-  <h3 class="section-head text-center">Kubeflow Components</h3>
   <div class="container">
+    <h3 class="section-head text-center">
```
I would suggest we whethere:
1. Remove this green lines between divs
2. Change it back to blue color.

I don't really like how this green line looks like on the white screen.

WDYT @thesuperzapper ?
![Screenshot 2025-03-30 at 00 39 18](https://github.com/user-attachments/assets/130c1169-3da2-4b2d-b3f7-31d1478e318a)

thesuperzapper replied:
Its technically a "teal" color, but we can make it dark/light blue if you want, let me see what it looks like.

thesuperzapper replied:
@andreyvelich I agree it looks much better with a slightly transparent grey (the darkness of which depends on if its dark/light mode).

Check it out now, and tell me if you like it (might need to do a "hard refresh and clear cache" by holding down on the refresh button while the dev tools panel is open).

- https://deploy-preview-3981--competent-brattain-de2d6d.netlify.app/

andreyvelich replied:
Yeah, looks great.

andreyvelich reviewed (commented):
Thank you for this effort @thesuperzapper!
Overall lgtm, just small comment.
/assign @kubeflow/release-team @kubeflow/kubeflow-steering-committee @varodrig @rimolive

terrytangyuan reviewed (commented):
Is KServe logo also updated?

thesuperzapper commented:
> Is KServe logo also updated?

Yes

andreyvelich commented:
Thank you @thesuperzapper!
/lgtm
/assign @franciscojavierarceo @terrytangyuan

terrytangyuan reviewed (approved):
/lgtm
/approve

franciscojavierarceo reviewed (approved):
Big thank you @thesuperzapper for doing this. Huge for the community!

andreyvelich commented:
thanks everyone.
/hold cancel

terrytangyuan commented:
@thesuperzapper One small suggestion: could we make it a toggle instead of a dropdown?

thesuperzapper commented:
> @thesuperzapper One small suggestion: could we make it a toggle instead of a dropdown?

We dont really have a pre-built toggle element, and I don't have a lot of time to make a new one.

If someone wants to propose a PR that changes it then we can discuss it there.

terrytangyuan commented:
No problem. Just wondering.
